### PR TITLE
RTL based on remaining flight time estimation

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by this pull request**
The current range-based RTL from https://github.com/PX4/PX4-Autopilot/pull/16399 needs to be configured with a flight time per battery which would need to be adjusted to the environment, flight speed, payload, ... every time it is used and results in very coarse estimate. It's IMO also less transparent for user feedback because the `rtl_limit_fraction` is not an intuitive graspable number.

**Describe your solution**
The approach I implemented in summary uses the remaining battery time from the battery estimation https://github.com/PX4/PX4-Autopilot/pull/17828 and compares it against a precise estimate how long the RTL will take. It triggers the RTL a configurable margin before that such that there's enough time to return before the battery is empty. This results in a different return time depending on the average current consumption and distance the vehicle is away from where it wants to land again e.g. home.

**Test data / coverage**
This strategy was tested a lot before but not in the ported form of this pr.

### CODE REVIEW

**Comment by MaEtUgR** [comment]
I rebased on top of master with the small conflict: https://github.com/PX4/PX4-Autopilot/pull/18645/files#diff-a58612b34e98d7c47e0b13294ccc0c5077dbab967450a7bc84acd596825a8151R86

File: msg/rtl_flight_time.msg [hunk 1]
```diff
@@ -1,4 +0,0 @@
-uint64 timestamp				# time since system start (microseconds)
-
-float32 rtl_time_s				# how long in seconds will the RTL take
-float32 rtl_limit_fraction			# what fraction of the allowable RTL time would be taken
```
Comment (RomanBapst) [inline msg/rtl_flight_time.msg:4]: @MaEtUgR Thanks for removing this, I always had trouble with this

File: src/modules/commander/Commander.cpp [hunk 2]
```diff
@@ -3523,77 +3525,69 @@ void Commander::avoidance_check()
...
+	    && _armed.armed
+	    && !_rtl_time_actions_done
+	    && PX4_ISFINITE(worst_battery_time_s)
+	    && rtl_time_estimate.safe_time_estimate >= worst_battery_time_s
+	    && _internal_state.main_state != commander_state_s::MAIN_STATE_AUTO_RTL
+	    && _internal_state.main_state != commander_state_s::MAIN_STATE_AUTO_LAND) {
+		// Try to trigger RTL
+		if (main_state_transition(_status, commander_state_s::MAIN_STATE_AUTO_RTL, _status_flags,
+					  _internal_state) == TRANSITION_CHANGED) {
+			mavlink_log_emergency(&_mavlink_log_pub, "Flight time low, returning to land");
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/commander/Commander.cpp:3584]:
```suggestion
			mavlink_log_emergency(&_mavlink_log_pub, "Remaining flight time low, returning to land");
```

[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/commander/Commander.cpp:3584]: I guess we should also add an event here and not only the legacy mavlink log?

File: src/modules/commander/Commander.cpp [hunk 3]
```diff
@@ -3523,77 +3525,69 @@ void Commander::avoidance_check()
...
 
-		} else if (battery_at_home < _param_bat_low_thr.get()) {
-			battery_range_warning = battery_status_s::BATTERY_WARNING_LOW;
+		} else {
+			mavlink_log_emergency(&_mavlink_log_pub, "Flight time low, land now!");
```
Comment (sfuhrer) [inline src/modules/commander/Commander.cpp:3587]:
```suggestion
			mavlink_log_emergency(&_mavlink_log_pub, "Remaining flight time low, land now!");
```

File: src/modules/navigator/rtl_params.c [hunk 4]
```diff
@@ -185,3 +174,32 @@ PARAM_DEFINE_FLOAT(RTL_LOITER_RAD, 50.0f);
...
+ * Time with margin = RTL_TIME_FACTOR * time + RTL_TIME_MARGIN
+ *
+ * @unit s
+ * @min 0
+ * @max 300
+ * @decimal 1
+ * @increment 1
+ * @group Return To Land
+ */
+PARAM_DEFINE_INT32(RTL_TIME_MARGIN, 110);
```
Comment (sfuhrer) [inline src/modules/navigator/rtl_params.c:205]: Not that it really matters, but it's an interesting default, why not making it 100s or 2min?

File: src/modules/navigator/rtl_params.c [hunk 5]
```diff
@@ -185,3 +174,32 @@ PARAM_DEFINE_FLOAT(RTL_LOITER_RAD, 50.0f);
...
+
+/**
+ * RTL time estimate safety margin offset
+ *
+ * Margin that is added to the time estimate, after it has already been scaled
+ * Time with margin = RTL_TIME_FACTOR * time + RTL_TIME_MARGIN
+ *
+ * @unit s
+ * @min 0
+ * @max 300
```
Comment (sfuhrer) [inline src/modules/navigator/rtl_params.c:200]: 300s sounds quite low as max, for vehicle flying multiple hours I would for sure want a bigger margin.
Would propose 3600 as max.

**Review by sfuhrer — APPROVED** [review]
I think it's fine to bring it in without major changes, though we should think about the following for the next iteration:
- I think it should be **either** the old battery state check triggering the warning/RTL/Land action, **or** this new range based RTL
- VTOL average current calculation needs some revising, probably it makes most sense if it only is updated in FW flight, and for the MC part of the flight the user sets reasonable safety margins. Or we add separate estimate for MC flight as well
- think about estimating power consumption or  an unitless "state or charge per second" instead of current estimation (current increases towards end of flight)

**Comment by RomanBapst** [comment]
@sfuhrer 

> VTOL average current calculation needs some revising, probably it makes most sense if it only is updated in FW flight, and for the MC part of the flight the user sets reasonable safety margins. Or we add separate estimate for MC flight as well

What issues do you currently see?

> think about estimating power consumption or an unitless "state or charge per second" instead of current estimation (current increases towards end of flight)

I think this could make sense. If we can get a reasonable "state of energy" estimate in [Wh] then the accuracy might be improved since it would be independent from current increase towards the end.

**Review by RomanBapst — APPROVED** [review]

**Comment by sfuhrer** [comment]
>What issues do you currently see?

Let's say the VTOL uses about 10x as much energy for hover as for cruise. If you update the consumption average estimation independently of the flight phase, then the remaining flight times drops by a factor of 10 when you transition to hover, likely triggering a RTL if the vehicle is at some distance to home.

But as I said, let's tackle that in a follow up.

File: src/modules/navigator/rtl_params.c (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/navigator/rtl_params.c:205]: Taken over from some old real use case. Not sure how it ended up being this value. A round number makes more sense as default indeed.

**Comment by MaEtUgR** [comment]
Thanks for the review! I'm quickly fixing the smaller issues @sfuhrer pointed out and then we can follow up with further improvements. I generally agree with all the points and am happy to discuss the details.

**Comment by MaEtUgR** [comment]
Or may I merge this and create an immediate follow-up for the fixes? That would help the work flow 😇

**Comment by sfuhrer** [comment]
> Or may I merge this and create an immediate follow-up for the fixes? That would help the work flow innocent

Whatever you prefer

File: src/modules/commander/Commander.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/commander/Commander.cpp:3584]: https://github.com/PX4/PX4-Autopilot/pull/18711

File: src/modules/navigator/rtl_params.c (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/navigator/rtl_params.c:205]: https://github.com/PX4/PX4-Autopilot/pull/18711

File: src/modules/navigator/rtl_params.c (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/navigator/rtl_params.c:200]: https://github.com/PX4/PX4-Autopilot/pull/18711

File: src/modules/commander/Commander.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/commander/Commander.cpp:3587]: https://github.com/PX4/PX4-Autopilot/pull/18711

**Comment by MaEtUgR** [comment]
I added the follow up here: https://github.com/PX4/PX4-Autopilot/pull/18711 Good input!

**Comment by hamishwillee** [comment]
@MaEtUgR What is the docs implication of this? 

I see `RTL_TIME_FACTOR` is new. I don't see any information about this in https://docs.px4.io/main/en/config/safety.html or http://localhost:8080/px4_user_guide/en/flight_modes/return.html

Specifically, what should a user set or not set to use this? It appears to be triggered of a low battery - what battery level? What if there are multiple batteries? How is it enabled/disabled?

**Pull request merged**